pyoozie/tags.py

```diff
@@ -505,35 +624,68 @@ def __init__(
             job_tracker=None,             # type: typing.Optional[typing.Text]
             name_node=None,               # type: typing.Optional[typing.Text]
             job_xml_files=None,           # type: typing.Optional[JobXmlFilesType]
+            actions=None,                 # type: typing.Optional[_WorkflowEntity]
     ):
         # type: (...) -> None
         XMLSerializable.__init__(self, 'workflow-app')
-        self.name = validate_xml_name(name)
 
-        self.parameters = Parameters(parameters)
-        self.global_configuration = GlobalConfiguration(
+        # XML-document-related values
+        self.__name = validate_xml_name(name)
+        self.__parameters = Parameters(parameters)
+        self.__global_configuration = GlobalConfiguration(
             job_tracker=job_tracker,
             name_node=name_node,
             job_xml_files=job_xml_files,
             configuration=configuration
         )
-        self.credentials = credentials
+        self.__credentials = credentials or []
+        self.__actions = actions
+        self.__validate()
+
+    def __validate(self):
+        # Parse actions for attributes
+        action_identifiers = []
+        credentials_needed = set()
+
+        def _parse_action(action):
+            action_identifiers.append(action.identifier())
+            if hasattr(action, 'credential'):
+                credential = action.credential()
+                if credential:
+                    credentials_needed.add(credential)
+
+        if self.__actions:
+            for action in set(self.__actions):
```

Comment:
In future PRs these iterators may return multiple instances of the same object since they are performing graph traversal and an action's OK and ERROR conditions could both point to the same object.